### PR TITLE
Experimental UdfMetric

### PR DESCRIPTION
**Pull request by richard-rogers**
## Description

Experimental call-back metric prompted by discussions on text support.
See soon-to-follow column augmenter PR for an alternative design.

## Changes

- Adds CallbackMetric and associated configuration

- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

File: python/whylogs/experimental/core/metrics/callback_metric.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    ):
+        super().__init__(dict())  # submetrics)
+        self._callbacks = callbacks
+        self._submetric_schema = submetric_schema or DefaultSchema()
+        self._type_mapper = type_mapper or StandardTypeMapper()
+        self._fi_disabled = fi_disabled
+
+    @property
+    def namespace(self) -> str:
+        return "callback"
```
Comment (richard-rogers) [inline python/whylogs/experimental/core/metrics/callback_metric.py]: consider making the namespace configurable

File: python/whylogs/experimental/core/metrics/callback_metric.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    def _update_relevant_submetrics(self, name: str, data: PreprocessedColumn) -> None:
+        if name not in self.submetrics:
+            return
+
+        for metric in self.submetrics[name].values():
+            metric.columnar_update(data)
+
+    def columnar_update(self, view: PreprocessedColumn) -> OperationResult:
+        count = 0
+        for value in list(chain.from_iterable(view.raw_iterator())):
```
Comment (richard-rogers) [inline python/whylogs/experimental/core/metrics/callback_metric.py]: special case for the different vies and map/apply where possible

File: python/whylogs/experimental/core/metrics/udf_metric.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        }
+
+        if isinstance(why_type, Integral):
+            metrics["distribution"] = StandardMetric.distribution.zero(MetricConfig())
+            metrics["ints"] = StandardMetric.ints.zero(MetricConfig())
+            if not fi_disabled:
+                metrics["frequent_items"] = StandardMetric.frequent_items.zero(MetricConfig())
+        elif isinstance(why_type, Fractional):
+            metrics["distribution"] = StandardMetric.distribution.zero(MetricConfig())
+        elif isinstance(why_type, String) and not fi_disabled:
```
Comment (jamie256) [inline python/whylogs/experimental/core/metrics/udf_metric.py:42]: consider this to make it easier to write metrics that cover string values on images.
```suggestion
        if not fi_disabled:
```

**Review by jamie256 — APPROVED** [review]
LGTM!

File: python/whylogs/experimental/core/metrics/callback_metric.py (hunk 1 above)
Comment (jamie256) [reply python/whylogs/experimental/core/metrics/callback_metric.py]: I think the 'udf' part is fine, and you get to name the udf function. We might strip it out when displaying in reports or in WhyLabs.

File: python/whylogs/experimental/core/metrics/callback_metric.py (hunk 2 above)
Comment (jamie256) [reply python/whylogs/experimental/core/metrics/callback_metric.py]: Good idea for follow up, we can start with this as is.

**Pull request merged**